<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=Generator content="Microsoft Word 15 (filtered)">
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:"MS Mincho";
	panose-1:2 2 6 9 4 2 5 8 3 4;}
@font-face
	{font-family:"MS Gothic";
	panose-1:2 11 6 9 7 2 5 8 2 4;}
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
@font-face
	{font-family:"Calibri Light";
	panose-1:2 15 3 2 2 2 4 3 2 4;}
@font-face
	{font-family:"\@MS Gothic";
	panose-1:2 11 6 9 7 2 5 8 2 4;}
@font-face
	{font-family:"Adobe Gothic Std B";
	panose-1:0 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:"Lucida Grande";}
@font-face
	{font-family:"\@Adobe Gothic Std B";
	panose-1:0 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:"\@MS Mincho";
	panose-1:2 2 6 9 4 2 5 8 3 4;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
h1
	{mso-style-link:"Heading 1 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	page-break-after:avoid;
	font-size:9.0pt;
	font-family:"Arial",sans-serif;
	font-weight:bold;
	text-decoration:underline;}
h2
	{mso-style-link:"Heading 2 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	page-break-after:avoid;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;
	font-style:italic;}
h3
	{mso-style-link:"Heading 3 Char";
	margin-top:0in;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	page-break-after:avoid;
	font-size:8.0pt;
	font-family:"Arial",sans-serif;
	font-weight:normal;
	text-decoration:underline;}
h4
	{mso-style-link:"Heading 4 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	page-break-after:avoid;
	font-size:14.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;}
h5
	{mso-style-link:"Heading 5 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:13.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;
	font-style:italic;}
h6
	{mso-style-link:"Heading 6 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:11.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;}
p.MsoHeading7, li.MsoHeading7, div.MsoHeading7
	{mso-style-link:"Heading 7 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoHeading8, li.MsoHeading8, div.MsoHeading8
	{mso-style-link:"Heading 8 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;
	font-style:italic;}
p.MsoHeading9, li.MsoHeading9, div.MsoHeading9
	{mso-style-link:"Heading 9 Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:11.0pt;
	font-family:"Calibri Light",sans-serif;}
p.MsoFootnoteText, li.MsoFootnoteText, div.MsoFootnoteText
	{mso-style-link:"Footnote Text Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:10.0pt;
	font-family:"Times New Roman",serif;}
p.MsoCommentText, li.MsoCommentText, div.MsoCommentText
	{mso-style-link:"Comment Text Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:10.0pt;
	font-family:"Times New Roman",serif;}
p.MsoHeader, li.MsoHeader, div.MsoHeader
	{mso-style-link:"Header Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoFooter, li.MsoFooter, div.MsoFooter
	{mso-style-link:"Footer Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
span.MsoFootnoteReference
	{vertical-align:super;}
p.MsoTitle, li.MsoTitle, div.MsoTitle
	{mso-style-link:"Title Char";
	margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:center;
	font-size:16.0pt;
	font-family:"Adobe Gothic Std B",sans-serif;
	font-weight:bold;}
p.MsoSubtitle, li.MsoSubtitle, div.MsoSubtitle
	{mso-style-link:"Subtitle Char";
	margin-top:0in;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:center;
	font-size:9.0pt;
	font-family:"Calibri Light",sans-serif;}
a:link, span.MsoHyperlink
	{color:#0563C1;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:#954F72;
	text-decoration:underline;}
em
	{font-family:"Calibri",sans-serif;
	font-weight:bold;}
p
	{margin-right:0in;
	margin-left:0in;
	font-size:12.0pt;
	font-family:"Times New Roman",serif;}
pre
	{mso-style-link:"HTML Preformatted Char";
	margin:0in;
	font-size:10.0pt;
	font-family:"Courier New";}
p.MsoCommentSubject, li.MsoCommentSubject, div.MsoCommentSubject
	{mso-style-link:"Comment Subject Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:10.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;}
p.MsoAcetate, li.MsoAcetate, div.MsoAcetate
	{mso-style-link:"Balloon Text Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Lucida Grande";}
span.MsoPlaceholderText
	{color:gray;}
p.MsoNoSpacing, li.MsoNoSpacing, div.MsoNoSpacing
	{margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoRMPane, li.MsoRMPane, div.MsoRMPane
	{margin:0in;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoListParagraph, li.MsoListParagraph, div.MsoListParagraph
	{margin-top:0in;
	margin-right:0in;
	margin-bottom:0in;
	margin-left:.5in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoListParagraphCxSpFirst, li.MsoListParagraphCxSpFirst, div.MsoListParagraphCxSpFirst
	{margin-top:0in;
	margin-right:0in;
	margin-bottom:0in;
	margin-left:.5in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoListParagraphCxSpMiddle, li.MsoListParagraphCxSpMiddle, div.MsoListParagraphCxSpMiddle
	{margin-top:0in;
	margin-right:0in;
	margin-bottom:0in;
	margin-left:.5in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoListParagraphCxSpLast, li.MsoListParagraphCxSpLast, div.MsoListParagraphCxSpLast
	{margin-top:0in;
	margin-right:0in;
	margin-bottom:0in;
	margin-left:.5in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoQuote, li.MsoQuote, div.MsoQuote
	{mso-style-link:"Quote Char";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;
	font-style:italic;}
p.MsoIntenseQuote, li.MsoIntenseQuote, div.MsoIntenseQuote
	{mso-style-link:"Intense Quote Char";
	margin-top:0in;
	margin-right:.5in;
	margin-bottom:0in;
	margin-left:.5in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;
	font-style:italic;}
span.MsoSubtleEmphasis
	{color:#5A5A5A;
	font-style:italic;}
span.MsoIntenseEmphasis
	{font-weight:bold;
	font-style:italic;
	text-decoration:underline;}
span.MsoSubtleReference
	{text-decoration:underline;}
span.MsoIntenseReference
	{font-weight:bold;
	text-decoration:underline;}
span.MsoBookTitle
	{font-family:"Calibri Light",sans-serif;
	font-weight:bold;
	font-style:italic;}
p.MsoBibliography, li.MsoBibliography, div.MsoBibliography
	{margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
p.MsoTocHeading, li.MsoTocHeading, div.MsoTocHeading
	{margin-top:12.0pt;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	page-break-after:avoid;
	font-size:9.0pt;
	font-family:"Arial",sans-serif;
	font-weight:bold;
	text-decoration:underline;}
span.TitleChar
	{mso-style-name:"Title Char";
	mso-style-link:Title;
	font-family:"Adobe Gothic Std B",sans-serif;
	font-weight:bold;}
span.Heading1Char
	{mso-style-name:"Heading 1 Char";
	mso-style-link:"Heading 1";
	font-family:"Arial",sans-serif;
	font-weight:bold;
	text-decoration:underline;}
span.Heading2Char
	{mso-style-name:"Heading 2 Char";
	mso-style-link:"Heading 2";
	font-family:"Times New Roman",serif;
	font-weight:bold;
	font-style:italic;}
span.Heading3Char
	{mso-style-name:"Heading 3 Char";
	mso-style-link:"Heading 3";
	font-family:"Arial",sans-serif;
	text-decoration:underline;}
span.Heading4Char
	{mso-style-name:"Heading 4 Char";
	mso-style-link:"Heading 4";
	font-weight:bold;}
span.Heading5Char
	{mso-style-name:"Heading 5 Char";
	mso-style-link:"Heading 5";
	font-weight:bold;
	font-style:italic;}
span.Heading6Char
	{mso-style-name:"Heading 6 Char";
	mso-style-link:"Heading 6";
	font-weight:bold;}
span.Heading7Char
	{mso-style-name:"Heading 7 Char";
	mso-style-link:"Heading 7";}
span.Heading8Char
	{mso-style-name:"Heading 8 Char";
	mso-style-link:"Heading 8";
	font-style:italic;}
span.Heading9Char
	{mso-style-name:"Heading 9 Char";
	mso-style-link:"Heading 9";
	font-family:"Calibri Light",sans-serif;}
span.SubtitleChar
	{mso-style-name:"Subtitle Char";
	mso-style-link:Subtitle;
	font-family:"Calibri Light",sans-serif;}
span.QuoteChar
	{mso-style-name:"Quote Char";
	mso-style-link:Quote;
	font-style:italic;}
span.IntenseQuoteChar
	{mso-style-name:"Intense Quote Char";
	mso-style-link:"Intense Quote";
	font-weight:bold;
	font-style:italic;}
span.HeaderChar
	{mso-style-name:"Header Char";
	mso-style-link:Header;
	font-family:"Times New Roman",serif;}
span.FooterChar
	{mso-style-name:"Footer Char";
	mso-style-link:Footer;
	font-family:"Times New Roman",serif;}
p.Keywords, li.Keywords, div.Keywords
	{mso-style-name:Keywords;
	mso-style-link:"Keywords Char";
	margin-top:0in;
	margin-right:0in;
	margin-bottom:12.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	border:none;
	padding:0in;
	font-size:9.0pt;
	font-family:"Times New Roman",serif;}
span.KeywordsChar
	{mso-style-name:"Keywords Char";
	mso-style-link:Keywords;
	font-family:"Times New Roman",serif;}
span.BalloonTextChar
	{mso-style-name:"Balloon Text Char";
	mso-style-link:"Balloon Text";
	font-family:"Lucida Grande";}
p.MaterialsandMethodsText, li.MaterialsandMethodsText, div.MaterialsandMethodsText
	{mso-style-name:"Materials and Methods Text";
	margin-top:0in;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	line-height:85%;
	font-size:8.0pt;
	font-family:"Arial",sans-serif;}
p.References, li.References, div.References
	{mso-style-name:References;
	margin-top:0in;
	margin-right:0in;
	margin-bottom:3.0pt;
	margin-left:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:8.0pt;
	font-family:"Arial",sans-serif;}
p.Authors, li.Authors, div.Authors
	{mso-style-name:Authors;
	margin:0in;
	text-align:center;
	font-size:10.0pt;
	font-family:"Times New Roman",serif;
	font-weight:bold;}
p.FigureCaption, li.FigureCaption, div.FigureCaption
	{mso-style-name:"Figure Caption";
	margin:0in;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:8.0pt;
	font-family:"Arial",sans-serif;}
span.HTMLPreformattedChar
	{mso-style-name:"HTML Preformatted Char";
	mso-style-link:"HTML Preformatted";
	font-family:"Courier New";}
span.kn
	{mso-style-name:kn;}
span.nn
	{mso-style-name:nn;}
span.k
	{mso-style-name:k;}
span.n
	{mso-style-name:n;}
span.o
	{mso-style-name:o;}
span.p
	{mso-style-name:p;}
span.s2
	{mso-style-name:s2;}
span.kc
	{mso-style-name:kc;}
span.nb
	{mso-style-name:nb;}
p.NotebookIn, li.NotebookIn, div.NotebookIn
	{mso-style-name:NotebookIn;
	mso-style-link:"NotebookIn Char";
	margin:0in;
	background:#F7F7F7;
	word-break:break-all;
	border:none;
	padding:0in;
	font-size:8.0pt;
	font-family:"Courier New";}
p.NotebookOut, li.NotebookOut, div.NotebookOut
	{mso-style-name:NotebookOut;
	mso-style-link:"NotebookOut Char";
	margin:0in;
	background:white;
	word-break:break-all;
	font-size:8.0pt;
	font-family:"Courier New";}
span.NotebookInChar
	{mso-style-name:"NotebookIn Char";
	mso-style-link:NotebookIn;
	font-family:"Courier New";
	background:#F7F7F7;}
span.mi
	{mso-style-name:mi;}
span.NotebookOutChar
	{mso-style-name:"NotebookOut Char";
	mso-style-link:NotebookOut;
	font-family:"Courier New";
	background:white;}
p.NotebookIn2, li.NotebookIn2, div.NotebookIn2
	{mso-style-name:NotebookIn2;
	mso-style-link:"NotebookIn2 Char";
	margin:0in;
	background:#F7F7F7;
	word-break:break-all;
	border:none;
	padding:0in;
	font-size:8.0pt;
	font-family:"Courier New";}
span.NotebookIn2Char
	{mso-style-name:"NotebookIn2 Char";
	mso-style-link:NotebookIn2;
	font-family:"Courier New";
	background:#F7F7F7;}
span.s1
	{mso-style-name:s1;}
span.CommentTextChar
	{mso-style-name:"Comment Text Char";
	mso-style-link:"Comment Text";
	font-family:"Times New Roman",serif;}
span.CommentSubjectChar
	{mso-style-name:"Comment Subject Char";
	mso-style-link:"Comment Subject";
	font-family:"Times New Roman",serif;
	font-weight:bold;}
span.pre
	{mso-style-name:pre;}
span.FootnoteTextChar
	{mso-style-name:"Footnote Text Char";
	mso-style-link:"Footnote Text";
	font-family:"Times New Roman",serif;}
.MsoChpDefault
	{font-family:"Calibri",sans-serif;}
 /* Page Definitions */
 @page WordSection1
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection1
	{page:WordSection1;}
@page WordSection2
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection2
	{page:WordSection2;}
@page WordSection3
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection3
	{page:WordSection3;}
@page WordSection4
	{size:8.5in 11.0in;
	margin:1.0in .5in 1.0in .5in;}
div.WordSection4
	{page:WordSection4;}
@page WordSection5
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection5
	{page:WordSection5;}
 /* List Definitions */
 ol
	{margin-bottom:0in;}
ul
	{margin-bottom:0in;}
-->
</style>

</head>

<body lang=EN-US link="#0563C1" vlink="#954F72" style='word-wrap:break-word'>

<div class=WordSection1>

<p class=MsoTitle>PySnpTools: <br>
Reading and Manipulating Genomic Data in Python </p>

</div>

<b><span style='font-size:16.0pt;font-family:"Adobe Gothic Std B",sans-serif'><br
clear=all style='page-break-before:auto'>
</span></b>

<div class=WordSection2>

<p class=MsoNormal>&nbsp;</p>

<p class=Authors>Carl M Kadie</p>

<p class=Authors><span style='font-weight:normal'>Microsoft Research (Retired)</span></p>

<p class=Authors><span style='font-weight:normal'>Redmond, WA 98052</span></p>

<p class=Authors><span style='font-family:"Calibri",sans-serif;font-weight:
normal'>carlk@msn.com</span></p>

</div>

<span style='font-size:10.0pt;font-family:"Times New Roman",serif'><br
clear=all style='page-break-before:auto'>
</span>

<div class=WordSection3>

<p class=MsoNormal>&nbsp;</p>

<h1>Abstract</h1>

<p class=MsoNormal align=left style='text-align:left'>PySnpTools is a Python
library of reading and manipulating genomic data in Python 3 (and Python 2). It
allows users to efficiently select and reorder individuals (rows) and SNP
locations (columns). It then reads only the data selected. Originally developed
to support FaST-LMM – a genome-wide-association-study (GWAS) tool – PySnpTools
now supports large-memory and cluster-scale work.</p>

<p class=MsoNormal align=left style='text-align:left'><br>
More generally, PySnpTools was inspired by NumPy and Pandas. It can be thought
of as a way to add Pandas-like indexing to 2&#8209;D NumPy arrays.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:none;border-bottom:solid black 1.0pt;padding:0in 0in 4.0pt 0in'>

<p class=Keywords align=left style='text-align:left'>Keywords: Python, Open
Source, Genomics, Scalability</p>

</div>

</div>

<span style='font-size:9.0pt;font-family:"Times New Roman",serif'><br
clear=all style='page-break-before:auto'>
</span>

<div class=WordSection4>

<h1>Introduction</h1>

<p class=MsoNormal>As we developed the FaST-LMM package (1;
2; 3; 4; 5; 6), we tired of re-writing our code to support more and more file formats
(<i>e.g.</i>, PLINK’s Bed, Dat, Pheno, etc.) (7; 8). Moreover, we noticed
ourselves repeatedly performing similar manipulations, for example,</p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>reading data for just a subset of SNPs (columns),</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>reordering the individuals (rows) in our phenotype data to match
their order in the SNP data, and </p>

<p class=MsoListParagraphCxSpLast style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>filling in missing data and normalizing data.</p>

<p class=MsoNormal>Inspired by NumPy and Pandas, we created PySnpTools, an
open-source library that makes these operations easy.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>We presented the first public version of PySnpTools at the
PyData 2015 conference in Seattle. At the conference, Travis Oliphant (primary
creator of NumPy) recommended PySnpTools’ approach to reading and manipulating
genomic data (9). Not long after the conference, Hilary Finucane (leader of the
Finucane Lab at the Broad Institute of MIT and Harvard) wrote us: “I've been
loving PySnpTools and recommending it to other statistical geneticists who work
in Python (10)!”</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Since the conference, as FaST-LMM grew to support datasets
of up to 1 million samples  (6), we expanded PySnpTools to also simplify:</p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>larger-than-memory datasets</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>running loops on multiple processors or on any clusters, and</p>

<p class=MsoListParagraphCxSpLast style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>reading and writing files locally or from/to any remote storage.</p>

<p class=MsoNormal>Most recently, we’ve added support for Python 3, a much-requested
feature.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>This paper tells how to install PySnpTools and describes the
genomic data that PySnpTools focuses on. It then gives examples of core usage.
It closes by listing other PySnpTools features and providing a brief comparison
to other dataset approaches.</p>

<h1 style='margin-bottom:6.0pt'>Installing PySnpTools</h1>

<p class=MsoNormal>To use PySnpTools:</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:.5in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>pip install pysnptools<a href="#_ftn1"
name="_ftnref1" title=""><span class=MsoFootnoteReference><span
class=MsoFootnoteReference><span style='font-size:8.0pt;font-family:"Courier New";
color:#333333'>[1]</span></span></span></a></span></p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Find Jupyter notebooks, full API documentation with examples,
and source code at <a href="https://github.com/fastlmm/PySnpTools">https://github.com/fastlmm/PySnpTools</a>.</p>

<h1 style='margin-bottom:6.0pt'>Genomic Data</h1>

<p class=MsoNormal>The genomic data of interest to us typically consists of
500,000 to 1.5 million columns -- one column for each SNP (that is a genome location
where humans are known to differ). The data includes of one row per individual.
One thousand to 1,000,000 rows are typical. Values within the initial data
might be 0,1,2 or missing (representing the number of minor alleles measured
for an individual at a genome location). After standardization, values are
64-bit or 32-bit floats, with missing values represented by NaN (“not-a-number”).</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Figure 1 shows PySnpTools’ in-memory representation of genomic
data. Two strings, called the <a
href="https://fastlmm.github.io/PySnpTools/#pysnptools.snpreader.SnpReader.iid"
title=pysnptools.snpreader.SnpReader.iid><span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New";color:windowtext;text-decoration:
none'>iid</span></b></span></a>, identify each individual. One string, called
the <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>sid</span></b></span>,
identifies each SNP. A float, called <span class=pre><b><span style='font-size:
8.0pt;font-family:"Courier New"'>val</span></b></span>, tells an individual’s allele
count at a SNP. Finally, a triple of floats, called <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>pos</span></b></span>, tells
the position of each SNP (chromosome number, genetic distance, and base-pair
position). For a given position in <span class=pre><b><span style='font-size:
8.0pt;font-family:"Courier New"'>val</span></b></span>, PySnpTools makes it
easy to find the corresponding <span class=pre><b><span style='font-size:8.0pt;
font-family:"Courier New"'>iid</span></b></span>, <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>sid</span></b></span>, and<span
class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'> pos.</span></b></span>
Moreover, for any <span class=pre><b><span style='font-size:8.0pt;font-family:
"Courier New"'>iids</span></b></span> or <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>sids</span></b></span> of
interest, PySnpTools makes it easy and efficient to find the corresponding
positions in <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>val</span></b></span>.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>For phenotypic and covariate data (<i>e.g.</i>, age, sex, height,
weight, presence of a disease), PySnpTools uses the same representation. For such
data, <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>sid</span></b></span>
tells the name of the feature (<i>e.g.</i>, “height”) while <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>pos</span></b></span> is
ignored. The <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>val</span></b></span><b>
</b>array, as a float, can also represent binary data via 0.0 and 1.0. We
handle categorical data via hot-one encoding.</p>

<h1 style='margin-bottom:6.0pt'><span style='position:relative;z-index:251662848;
left:-2px;top:0px;width:345px;height:272px'><img width=345 height=272
src="KadiePySnpTools_files/image001.png"></span><br clear=ALL>
Reading Genomic Files: An Example</h1>

<p class=MsoNormal>A Python object that reads genomic data from a file is
called a SnpReader. Here is how to create a SnpReader for the popular Bed file
format: We tell it what file to read and how to read it. (This SnpReader, as
yet reads nothing from its file.)</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#F7F7F7'>

<p class=NotebookIn><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></p>

<p class=NotebookIn><b><span style='color:green'>from</span><span
style='color:#333333'> </span><span style='color:black'>pysnptools.snpreader</span><span
style='color:#333333'> </span><span style='color:green'>import</span><span
style='color:#333333'> Bed</span></b></p>

<p class=NotebookIn><span style='color:#333333'>snpreader </span><span
style='color:#666666'>=</span><span style='color:#333333'> Bed(</span><span
style='color:#BA2121'>&quot;all.bed&quot;</span><span style='color:#333333'>,
count_A1</span><span style='color:#666666'>=</span><span style='color:green'>True</span><span
style='color:#333333'>)</span></p>

<p class=NotebookIn><span style='color:green'>print</span><span
style='color:#333333'>(snpreader)</span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<p class=NotebookOut><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Output:<br>
</span></i><span style='color:black'>Bed('all.bed',count_A1=True)</span></p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>We can ask a SnpReader for the number of individuals and the
number of SNPs. We can also, for example, ask it for the <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>iid</span></b></span> of the
first individual. We can additionally ask it for the <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>sid</span></b></span> of the
last SNP. (The Bed SnpReader reads only the small files needed to answer these
questions.)</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#F7F7F7'>

<p class=NotebookIn><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></p>

<p class=NotebookIn2><span class=nb><span style='color:green'>print</span></span><span
class=p><span style='color:#333333'>(</span></span><span class=n><span
style='color:#333333'>snpreader</span></span><span class=o><span
style='color:#666666'>.</span></span><span class=n><span style='color:#333333'>iid_count</span></span><span
class=p><span style='color:#333333'>,</span></span><span class=n><span
style='color:#333333'>snpreader</span></span><span class=o><span
style='color:#666666'>.</span></span><span class=n><span style='color:#333333'>sid_count</span></span><span
class=p><span style='color:#333333'>)</span></span></p>

<p class=NotebookIn2><span class=nb><span style='color:green'>print</span></span><span
class=p><span style='color:#333333'>(</span></span><span class=n><span
style='color:#333333'>snpreader</span></span><span class=o><span
style='color:#666666'>.</span></span><span class=n><span style='color:#333333'>iid</span></span><span
class=p><span style='color:#333333'>[</span></span><span class=mi><span
style='color:#666666'>0</span></span><span class=p><span style='color:#333333'>])</span></span></p>

<p class=NotebookIn2><span class=nb><span style='color:green'>print</span></span><span
class=p><span style='color:#333333'>(</span></span><span class=n><span
style='color:#333333'>snpreader</span></span><span class=o><span
style='color:#666666'>.</span></span><span class=n><span style='color:#333333'>sid</span></span><span
class=p><span style='color:#333333'>[</span></span><span class=o><span
style='color:#666666'>-</span></span><span class=mi><span style='color:#666666'>1</span></span><span
class=p><span style='color:#333333'>])</span></span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>500 5000</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>['cid0P0' 'cid0P0']</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>snp124_m0_.23m1_.08</span></pre>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Next, we can read all the genomic data into memory, creating
a new SnpReader called a SnpData. Because a SnpData is a SnpReader, we can
again ask for the number of individuals and SNPs.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#F7F7F7'>

<p class=NotebookIn><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></p>

<pre style='background:#F7F7F7;word-break:break-all;border:none;padding:0in'><span
class=n><span style='font-size:8.0pt;color:#333333'>snpdata</span></span><span
style='font-size:8.0pt;color:#333333'> </span><span class=o><span
style='font-size:8.0pt;color:#666666'>=</span></span><span style='font-size:
8.0pt;color:#333333'> </span><span class=n><span style='font-size:8.0pt;
color:#333333'>snpreader</span></span><span class=o><span style='font-size:
8.0pt;color:#666666'>.</span></span><span class=n><span style='font-size:8.0pt;
color:#333333'>read</span></span><span class=p><span style='font-size:8.0pt;
color:#333333'>()</span></span></pre><pre style='background:#F7F7F7;word-break:
break-all;border:none;padding:0in'><span class=nb><span style='font-size:8.0pt;
color:green'>print</span></span><span class=p><span style='font-size:8.0pt;
color:#333333'>(</span></span><span class=n><span style='font-size:8.0pt;
color:#333333'>snpdata</span></span><span class=p><span style='font-size:8.0pt;
color:#333333'>)</span></span></pre><pre style='background:#F7F7F7;word-break:
break-all;border:none;padding:0in'><span class=nb><span style='font-size:8.0pt;
color:green'>print</span></span><span class=p><span style='font-size:8.0pt;
color:#333333'>(</span></span><span class=n><span style='font-size:8.0pt;
color:#333333'>snpdata</span></span><span class=o><span style='font-size:8.0pt;
color:#666666'>.</span></span><span class=n><span style='font-size:8.0pt;
color:#333333'>iid_count</span></span><span class=p><span style='font-size:
8.0pt;color:#333333'>,</span></span><span style='font-size:8.0pt;color:#333333'> </span><span
class=n><span style='font-size:8.0pt;color:#333333'>snpdata</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>sid_count</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>)</span></span></pre></div>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>&nbsp;</span></pre><pre style='background:
white;vertical-align:baseline;word-break:break-all'><i><span style='font-size:
6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>SnpData(Bed('all.bed',count_A1=True))</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>500 5000</span></pre>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>A SnpData is a SnpReader that contains a <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>val</span></b></span> property
that other SnpReaders do not. The <span class=pre><b><span style='font-size:
8.0pt;font-family:"Courier New"'>val</span></b></span> property is an (in-memory)
NumPy array of the genomic data.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>We can, for example, show the genomic data for the first 7
individuals and first 7 SNPs. We can also find the mean of <b>all</b> the
genomic data.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 1.0pt 1.0pt 1.0pt;
background:#F7F7F7'><pre style='background:#F7F7F7;word-break:break-all;
border:none;padding:0in'><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></pre><pre style='background:#F7F7F7;word-break:
break-all;border:none;padding:0in'><span class=kn><b><span style='font-size:
8.0pt;color:green'>import</span></b></span><span style='font-size:8.0pt;
color:#333333'> </span><span class=nn><b><span style='font-size:8.0pt;
color:blue'>numpy</span></b></span><span style='font-size:8.0pt;color:#333333'> </span><span
class=k><b><span style='font-size:8.0pt;color:green'>as</span></b></span><span
style='font-size:8.0pt;color:#333333'> </span><span class=nn><b><span
style='font-size:8.0pt;color:blue'>np</span></b></span></pre><pre
style='background:#F7F7F7;word-break:break-all;border:none;padding:0in'><span
class=nb><span style='font-size:8.0pt;color:green'>print</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>(</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>snpdata</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>val</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>[:</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>7</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,:</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>7</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>])</span></span></pre><pre
style='background:#F7F7F7;word-break:break-all;border:none;padding:0in'><span
class=nb><span style='font-size:8.0pt;color:green'>print</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>(</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>np</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>mean</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>(</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>snpdata</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>val</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>))</span></span></pre></div>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>&nbsp;</span></pre><pre style='background:
white;vertical-align:baseline;word-break:break-all'><i><span style='font-size:
6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'> [[0. 0. 1. 2. 0. 1. 2.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 1. 1. 0. 0. 2.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 1. 2. 1. 0. 0.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 0. 2. 0. 0. 0.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 0. 0. 0. 0. 2.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 1. 0. 0. 0. 2.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 0. 2. 1. 0. 1. 2.]]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>0.521412</span></pre>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Alternatively, we can do everything in one line.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#F7F7F7'>

<p class=NotebookIn><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:green'>print</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>(np</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#666666'>.</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>mean(Bed(</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#BA2121'>&quot;all.bed&quot;</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>,count_A1</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>=</span><b><span
style='font-size:7.5pt;font-family:"Courier New";color:green'>True</span></b><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>)</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>read()</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>val))</span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><i><span style='font-size:6.0pt;
font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;font-family:"Courier New";color:black'>0.521412</span></p>

<h1 style='margin-bottom:6.0pt'>SnpData</h1>

<p class=MsoNormal>SnpData, the special in-memory SnpReader, is created by any
SnpReader's <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>read()</span></b></span>
method. We can also create a SnpData from scratch.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Here we create <span class=pre><b><span style='font-size:
8.0pt;font-family:"Courier New"'>snpdata1</span></b></span> for three
individuals and two SNPs. We use NaN to mark a missing value and then ask for
the mean value (ignoring the missing value).</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#F7F7F7'>

<p class=NotebookIn><i><span style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;
color:black'>Input:</span></i></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><b><span style='font-size:8.0pt;
font-family:"Courier New";color:green'>from</span></b><span style='font-size:
8.0pt;font-family:"Courier New";color:#333333'> </span><b><span
style='font-size:8.0pt;font-family:"Courier New";color:blue'>pysnptools.snpreader</span></b><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'> </span><b><span
style='font-size:8.0pt;font-family:"Courier New";color:green'>import</span></b><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'> SnpData</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>snpdata1 </span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'> SnpData(iid</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>[[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'f1'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'c1'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>],[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'f1'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'c2'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>],</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>                        [</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'f2'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'c1'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>]],</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>                   sid</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'snp1'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>'snp2'</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>],</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>                   val</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>[[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>0</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>1</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>],[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>2</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.5</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>],[</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.5</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,np</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>nan]])</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:green'>print</span><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>(np</span><span style='font-size:8.0pt;
font-family:"Courier New";color:#666666'>.</span><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>nanmean(snpdata1</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>val))</span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>0.8</span></pre>

<h1 style='margin-bottom:6.0pt'>Selecting and Reordering Data <i>Before</i>
Reading</h1>

<p class=MsoNormal>Suppose we only care about the genomic data for the first 7
individuals and first 7 SNPs. PySnpTools makes it easy to read just the desired
data from disk. We use NumPy-like indexing <i>before</i> the read method.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 1.0pt 1.0pt 1.0pt;
background:#F7F7F7'>

<p class=NotebookIn style='background:#F7F7F7;border:none;padding:0in'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Input:</span></i></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>snpreader </span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'> Bed(</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#BA2121'>&quot;all.bed&quot;</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,count_A1</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><b><span
style='font-size:8.0pt;font-family:"Courier New";color:green'>True</span></b><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>)</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>snpdata77 </span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'> snpreader[:</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>7</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>,:</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>7</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>]</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>read()</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:8.0pt;
font-family:"Courier New";color:green'>print</span><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>(snpdata77</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:8.0pt;font-family:"Courier New";color:#333333'>val)</span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>[[0. 0. 1. 2. 0. 1. 2.]</span></pre>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 1. 1. 0. 0. 2.]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 1. 2. 1. 0. 0.]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 0. 2. 0. 0. 0.]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 0. 0. 0. 0. 2.]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 1. 0. 0. 0. 2.]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:white;
vertical-align:baseline;word-break:break-all'><span style='font-size:8.0pt;
font-family:"Courier New";color:black'> [0. 0. 2. 1. 0. 1. 2.]]</span></p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'>&nbsp;</span></pre>

<p class=MsoNormal>All NumPy-like fancy indexing is supported: slicing, Booleans,
lists of integers, negative integers (11). Additionally, PySnpTools allows fancy
indexing on both rows and columns at once.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Appending indexing to any SnpReader, creates a new
SnpReader. Here is an extreme example. It says, “create a reader</p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>from file &quot;all.bed&quot; in Bed format</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>with the individuals in reverse order and for every 2nd SNP</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>for the first 5 such individuals and first 5 such SNPs</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>for the first and last individuals and the first, 2nd, and 5th
SNP,</p>

<p class=MsoListParagraphCxSpLast style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>then read.”</p>

<p class=MsoNormal>As before, only the final, desired data is read from the
disk.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 1.0pt 1.0pt 1.0pt;
background:#F7F7F7'>

<p class=NotebookIn style='background:#F7F7F7;border:none;padding:0in'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Input:</span></i></p>

<pre style='background:#F7F7F7;word-break:break-all;border:none;padding:0in'><span
class=nb><span style='font-size:8.0pt;color:green'>print</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>(</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>Bed</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>(</span></span><span
class=s2><span style='font-size:8.0pt;color:#BA2121'>&quot;all.bed&quot;</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>count_A1</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>=</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>True</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>)[::</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>-</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>1</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,::</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>2</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>][:</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>5</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,:</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>5</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>]<br>
       [[</span></span><span class=mi><span style='font-size:8.0pt;color:#666666'>0</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>-</span></span><span
class=mi><span style='font-size:8.0pt;color:#666666'>1</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>],[</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>True</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>True</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>False</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>False</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>,</span></span><span
class=kc><b><span style='font-size:8.0pt;color:green'>True</span></b></span><span
class=p><span style='font-size:8.0pt;color:#333333'>]]<br>
       </span></span><span class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>read</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>()</span></span><span
class=o><span style='font-size:8.0pt;color:#666666'>.</span></span><span
class=n><span style='font-size:8.0pt;color:#333333'>val</span></span><span
class=p><span style='font-size:8.0pt;color:#333333'>)</span></span></pre></div>

<p class=MsoNormal><span style='font-size:8.0pt'>&nbsp;</span></p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>[[0. 0. 1.]</span></pre><pre
style='background:white;vertical-align:baseline;word-break:break-all'><span
style='font-size:8.0pt;color:black'> [0. 1. 1.]]</span></pre>

<p class=MsoNormal>&nbsp;</p>

<h1 style='margin-bottom:6.0pt'>Indexing by Individual (Row) and SNP (Column) Identifiers
</h1>

<p class=MsoNormal>We've seen how to manipulate SnpReaders with position-based
indexing. What if we instead wish to manipulate SnpReaders by <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>iid</span></b></span> (the
individuals’ identifier) or <span class=pre><b><span style='font-size:8.0pt;
font-family:"Courier New"'>sid</span></b></span> (the SNPs’ identifier)? This
is a more Pandas-like scenario. In that case, we can use the <span class=pre><b><span
style='font-size:8.0pt;font-family:"Courier New"'>iid_to_index()</span></b></span>
or <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>sid_to_index()</span></b></span>
methods.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>This example shows how to read three SNPs of interest, each
identified by <b>sid</b> rather than position.</p>

<p class=MsoNormal>&nbsp;</p>

<div style='border:solid windowtext 1.0pt;padding:1.0pt 1.0pt 1.0pt 1.0pt;
background:#F7F7F7'>

<p class=NotebookIn style='background:#F7F7F7;border:none;padding:0in'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Input:</span></i></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>desired_sid_list </span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'> [</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#BA2121'>'snp1750_m0_.02m1_.04'</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>,<br>
             </span><span style='font-size:7.5pt;font-family:"Courier New";
color:#BA2121'>'snp0_m0_.37m1_.24'</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>,</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#BA2121'>'snp122_m0_.26m1_.34'</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>snpreader </span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'> Bed(</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#BA2121'>&quot;all.bed&quot;</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>,count_A1</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>=</span><b><span
style='font-size:7.5pt;font-family:"Courier New";color:green'>True</span></b><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>)</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>desired_snpreader </span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>=</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'> \</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>    snpreader[:,snpreader</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>sid_to_index(desired_sid_list)]</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:green'>print</span><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>(desired_snpreader</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>iid_count,</span></p>

<p class=MsoNormal align=left style='text-align:left;background:#F7F7F7;
word-break:break-all;border:none;padding:0in'><span style='font-size:7.5pt;
font-family:"Courier New";color:#333333'>      desired_snpreader</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#666666'>.</span><span
style='font-size:7.5pt;font-family:"Courier New";color:#333333'>sid_count)</span></p>

</div>

<p class=MsoNormal>&nbsp;</p>

<pre style='background:white;vertical-align:baseline;word-break:break-all'><i><span
style='font-size:6.0pt;font-family:"Calibri Light",sans-serif;color:black'>Output:<br>
</span></i><span style='font-size:8.0pt;color:black'>500 3</span></pre>

<p class=MsoNormal>&nbsp;</p>

<h1 style='margin-bottom:6.0pt'>Readers and Pseudo-Readers</h1>

<p class=MsoNormal>PySnpTools includes SnpReaders for these common file formats:
Bed, Pheno, Dat, Ped, Dense. It also defines these new binary formats: SnpNpz,
SnpHdf5, and SnpMemMap. The last is of interest because it uses NumPy’s
memory-mapped arrays to offer in-memory-like access to data larger than will
fit in memory. PySnpTools also defines a format called DistributedBed that
offers random access to data stored compactly across dozens (or hundreds or
thousands) of Bed files.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>In addition, PySnpTools includes SnpGen, which could be
called a “pseudo-reader”. To any program using SnpGen, it looks like a file
reader, but instead of reading data from the disk, it generates random genomic data
on the fly. This generation, based on a user-provided seed, is deterministic. </p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Beyond reading SNPS, PySnpTools includes a set of readers
and pseudo-readers for kernel data. (Kernel data represents the pair-wise similarity
between individuals.)</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>If PySnpTools doesn’t support a data format of interest, you
have two options. If your data is relatively small, you can just read it with
other Python tools and then create an in-memory SnpData (or SnpMemMap) object.
If more performance is required, you can write a new SnpReader module for the
format. In either case, any programs written to expect a SnpReader will
automatically work with the new data format.</p>

<h1 style='margin-bottom:6.0pt'>Beyond Genomic Data</h1>

<p class=MsoNormal>PySnpTools includes PstReader, a generalized class that adds
indexing-by-row-or-column-identifier to any 2&#8209;D numeric data. It also
adds row and column properties. Unlike SnpReader, it does not require rows to
be individuals or columns to be SNPs.</p>

<h1 style='margin-bottom:6.0pt'>Beyond Reading</h1>

<p class=MsoNormal>Recently, as we worked to scale FaST-LMM to 1 million
samples, we added more functionality to PySnpTools. PySnpTools now includes tools
for:</p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>standardizing SNP data and kernels,</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>intersecting (and ordering) the individuals from any number of
SnpReaders (in one line),</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>efficiently working with ranges of large integers,</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>writing loops so they can run on multiple processor or on any
cluster (defined by a module), and</p>

<p class=MsoListParagraphCxSpLast style='text-indent:-.25in'><span
style='font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span>reading and writing files to any local or any distributed file
system (defined by a module).</p>

<p class=MsoNormal>&nbsp;</p>

<h1 style='margin-bottom:6.0pt'>Comparing to Other Dataset Formats</h1>

<p class=MsoNormal>At PyData 2015, Joshua Bloom compared dataset tools for Data
Science (12). Table 1 updates his table with PySnpTools. Fundamentally, PySnpTools
decorates 2&#8209;D NumPy arrays with immutable, indexed row and column
identifiers. This gives us convenience and high efficiency but is more special
purpose than other approaches. PySnpTools also delays all copying (and reading)
until the <span class=pre><b><span style='font-size:8.0pt;font-family:"Courier New"'>read()</span></b></span>
method is explicitly called, which gives us efficiency and predictability.</p>

<h1 style='margin-bottom:6.0pt'>Summary</h1>

<p class=MsoNormal>We have developed PySnpTools, a Python library for reading
and manipulating genomic data. Inspired by NumPy and Pandas, PySnpTools can be
thought of as a way to add Pandas-like indexing to 2&#8209;D NumPy arrays. To install
PySnpTools:</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:.5in'><span style='font-size:8.0pt;
font-family:"Courier New";color:#333333'>pip install pysnptools</span></p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>Jupyter notebooks, full API documentation with examples, and
source code are available at <a href="https://github.com/fastlmm/PySnpTools">https://github.com/fastlmm/PySnpTools</a>.</p>

<h1 style='margin-bottom:6.0pt'><img width=724 height=248
src="KadiePySnpTools_files/image002.png" align=left hspace=12 vspace=5
alt="	Slicing Induces Copy	Immutable Columns	Query	Transfer Speed To Python	C++ SDK	Distributed	Memory Efficiency	Categorical Optimized	Sparse &amp; Dense&#13;&#10;PySnpTools	No	Yes &#13;&#10;(and Row)	Yes	NumPy	NumPy	No	NumPy	No	Dense Only&#13;&#10;Pandas DataFrame	Sequences	No	Yes	N/A	No	No	Medium	Medium	Yes&#13;&#10;GraphLab&#13;&#10;SFrame	Yes	Yes	Yes	Low	Yes	Yes	High	No	Yes&#13;&#10;Spark&#13;&#10;DataFrame	Yes	Yes	Yes	Very Low	No	Yes	Low	No	Yes&#13;&#10;Dask	Yes	No	Yes	N/A	No	Yes	Medium	No	No&#13;&#10;Blaze	No	No	Yes	N/A	No	Yes	Medium	No	No&#13;&#10;Wise DataSet	Copy-on-write	No	Yes	Very High	Yes	Yes	Very High	High	Yes&#13;&#10;Table 1. Bloom’s Dataset for Data Science, plus PySnpTools&#13;&#10;">Acknowledgments</h1>

<p class=MsoNormal>Thanks to everyone on the FaST-LMM team, whose use and feedback
informed the PySnpTools design. A special thanks to Christoph Lippert, who
wrote our original Bed reader, and David Heckerman, who encouraged PySnpTools’
development.</p>

<h1>References</h1>

<p class=MsoBibliography>1. <i>FaST linear mixed models for genome-wide
association studies. </i><b>Lippert, C.,<i> et al.</i></b> 2011, Nature
Methods, 8 833-835.</p>

<p class=MsoBibliography>2. <i>An Exhaustive Epistatic SNP Association Analysis
on Expanded Wellcome Trust Data. </i><b>Lippert, C., et al.</b> 2013,
Scientific Reports 3, 1099.</p>

<p class=MsoBibliography>3. <i>Greater power and computational efficiency for
kernel-based association testing of sets of genetic variants. </i><b>Lippert,
C, <i>et al.</i></b> 2014, Bioinformatics 30,22.</p>

<p class=MsoBibliography>4. <i>Further Improvements to Linear Mixed Models for
Genome-Wide Association Studies. </i><b>Widmer, C., <i>et al.</i></b> 2015,
Scientific Reports, 4 6874.</p>

<p class=MsoBibliography>5. <i>Linear mixed model for heritability estimation. </i><b>Heckerman,
D.<i> et al</i>.</b> 2016, Proceedings of the National Academy of Sciences 113
(27) .</p>

<p class=MsoBibliography>6. <i>Ludicrous Speed Linear Mixed Models for
Genome-Wide Association Studies. </i><b>Kadie, C. &amp; Heckerma, D.</b> 2019,
bioRxiv 154682.</p>

<p class=MsoBibliography>7. <b>Purcell, S.</b> PLINK. [Online]
http://zzz.bwh.harvard.edu/plink/.</p>

<p class=MsoBibliography>8. <i>PLINK: a toolset for whole-genome association
and population-based. </i><b>Purcell, S.<i> et al.</i></b> 2007, American
Journal of Human Genetics, 81.</p>

<p class=MsoBibliography>9. <b>Oliphant, T.</b> Personal Communications. 2015.</p>

<p class=MsoBibliography>10. <b>Finuncane, H.</b> Personal Communications.
2015.</p>

<p class=MsoBibliography align=left style='text-align:left'>11. <b>Indexing. <i>NumPy
User Guide. </i>[Online]
https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html.</b></p>

<p class=MsoBibliography><b>12. <i>A Systems View of Machine Learning. </i>Bloom,
J. Seattle&nbsp;: s.n., 2015. PyData.</b></p>

<p class=References><br clear=all>
</p>

</div>

<span style='font-size:8.0pt;font-family:"Arial",sans-serif'><br clear=all
style='page-break-before:auto'>
</span>

<div class=WordSection5>

<p class=References>&nbsp;</p>

</div>

<div><br clear=all>

<hr align=left size=1 width="33%">

<div id=ftn1>

<p class=FigureCaption><a href="#_ftnref1" name="_ftn1" title=""><span
class=MsoFootnoteReference><span class=MsoFootnoteReference><span
style='font-size:8.0pt;font-family:"Arial",sans-serif'>[1]</span></span></span></a>
Assuming pip version 10+ or Anaconda.</p>

</div>

</div>

</body>

</html>
